### PR TITLE
Add BufferView::section() method

### PR DESCRIPTION
**Pull request by laian**
I don't know if throwing a string is the best, so let me know what to change it to if I should change it!

@mbojrab @jmrandol

### CODE REVIEW

File: modules/c++/mem/include/mem/BufferView.h [hunk 1]
```diff
@@ -50,21 +51,52 @@ struct BufferView
...
+mem::BufferView<T> mem::BufferView<T>::section(size_t sectionSize)
+{
+    if(size < sectionSize)
+    {
+        throw std::string("section tried to take more space than available");
+    } 
+
+    mem::BufferView<T> newSection(data, sectionSize);
+
+    size -= size;
```
Comment (asylvest) [inline modules/c++/mem/include/mem/BufferView.h]: These should be offset by sectionSize, not size

File: modules/c++/mem/include/mem/BufferView.h [hunk 2]
```diff
@@ -50,21 +51,52 @@ struct BufferView
...
+
 };
+
+
+template <typename T>
+mem::BufferView<T> mem::BufferView<T>::section(size_t sectionSize)
+{
+    if(size < sectionSize)
+    {
+        throw std::string("section tried to take more space than available");
```
Comment (asylvest) [inline modules/c++/mem/include/mem/BufferView.h]: Throw an `except::Exception(Ctxt("Blah blah"))`.  Expand it to tell you how much space you had and how much you asked for.  Include `<sys/Conf.h>` and `except/Exception.h` to get this stuff.

File: modules/c++/mem/include/mem/BufferView.h [hunk 3]
```diff
@@ -50,21 +51,52 @@ struct BufferView
...
+
+    mem::BufferView<T> newSection(data, sectionSize);
+
+    size -= size;
+    data += size;
+
+    return newSection;
+}
+
+template <typename T>
```
Comment (asylvest) [inline modules/c++/mem/include/mem/BufferView.h]: For now at least keep the implementation of these two functions in the body of the class.  I don't disagree with you though that the code's easier to read when it's down there so we can start re-thinking our style guide sometime soon.

**Comment by laian** [comment]
The above suggestions have been fixed.

File: modules/c++/mem/include/mem/BufferView.h [hunk 4]
```diff
@@ -50,6 +53,34 @@ struct BufferView
     T* data;
     size_t size;
 
+    // returns a new bufferView that "takes" sectionSize T from the bufferView
+    // For instance, if the current bufferView had 100 ints, and you called 
+    // section with sectionSize = 10, section() would return a new BufferView
+    // of size 10 and the original BufferView would be left with 90 ints.
+    // throws std::string if there is not enough space
```
Comment (asylvest) [inline modules/c++/mem/include/mem/BufferView.h]: Update comment to reflect what type is thrown

Also, use Doxygen-style comments (I will admit we have often been lazy about this).  So use stuff like ```\throws``` to describe the exception

File: modules/c++/mem/include/mem/BufferView.h [hunk 5]
```diff
@@ -50,6 +53,34 @@ struct BufferView
...
+    // returns a new bufferView that "takes" sectionSize T from the bufferView
+    // For instance, if the current bufferView had 100 ints, and you called 
+    // section with sectionSize = 10, section() would return a new BufferView
+    // of size 10 and the original BufferView would be left with 90 ints.
+    // throws std::string if there is not enough space 
+    mem::BufferView<T> section(size_t sectionSize)
+    {
+        if(size < sectionSize)
+        {   
+            std::stringstream ss;
```
Comment (asylvest) [inline modules/c++/mem/include/mem/BufferView.h]: Use ostringstream rather than just stringstream

**Comment by laian** [comment]
Is there a good link for the commenting style?  I didn't see anything about putting /throws in the comments

**Comment by laian** [comment]
I gave my best attempt :dizzy_face:

**Pull request merged**